We wanted to check the runtime for the ketu code in our configuration.

In [1]:
import swarced as sw
import time, pickle, random, os, run
import numpy as np
import multiprocessing as mp

In [2]:
qlist = os.listdir("./timing")

#No parallel 

In [63]:
qlist = os.listdir("./timing/")
t = np.array([])
for q in random.sample(qlist,5):
    start = time.time()
    os.system("python run.py " + q[7:-6] + " 0 /k2_data/query/" + q)
    t = np.append(t,time.time()-start)

Without parallel processing each cycle took about 490 seconds with a standard deviation in a sample of 5 of 10 seconds. That means for 5 tasks it took about 40 minutes.

In [65]:
np.mean(t)

490.52086238861085

In [69]:
np.std(t)

9.9701895198145252

In [71]:
np.sum(t)/60.

40.876738532384238

#Multiprocessing 

##MAC: Nicole

In [9]:
import multiprocessing as mp

In [3]:
def single_run(qfile):
    start = time.time()
    #python run.py epicID campaign query_file_path
    #os.system("python run.py " + qfile[4:-2] + " 0 /k2_data/query/" + qfile)
    run.main([qfile[7:-6], 0, "/Users/mhughes/swarced/timing/" + qfile])
    return time.time() - start

In [8]:
for q in qlist:
    qdict = pickle.load(open('./timing/'+q,'r'))
    qdict['light_curve_file'] = '/Users/mhughes/swarced/lcs/ktwo' + str(q[7:-6]) + "-c00_lpd-lc.fits"
    qdict['catalog_file'] = '/Users/mhughes/Desktop/epic.h5'
    qdict['basis_file'] = '/Users/mhughes/Desktop/c0.h5'
    #qdict['basis_file'] = '/Users/mhughes/Desktop/c0.h5'
    sw.save_query(qdict, open("./timing/" + q, 'wb'))
    print qdict 

{'time_spacing': 0.1, 'durations': [0.05, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0], 'max_period': 70.0, 'nbasis': 150, 'light_curve_file': '/Users/mhughes/swarced/lcs/ktwo202060506-c00_lpd-lc.fits', 'initial_time': 1940.0, 'min_period': 0.5, 'basis_file': '/Users/mhughes/Desktop/c0.h5', 'catalog_file': '/Users/mhughes/Desktop/epic.h5'}
{'time_spacing': 0.1, 'durations': [0.05, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0], 'max_period': 70.0, 'nbasis': 150, 'light_curve_file': '/Users/mhughes/swarced/lcs/ktwo202060911-c00_lpd-lc.fits', 'initial_time': 1940.0, 'min_period': 0.5, 'basis_file': '/Users/mhughes/Desktop/c0.h5', 'catalog_file': '/Users/mhughes/Desktop/epic.h5'}
{'time_spacing': 0.1, 'durations': [0.05, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0], 'max_period': 70.0, 'nbasis': 150, 'light_curve_file': '/Users/mhughes/swarced/lcs/ktwo202071579-c00_lpd-lc.fits', 'initial_time': 1940.0, 'min_period': 0.5, 'basis_file': '/Users/mhughes/Desktop/c0.h5', 'catalog_

In [9]:
ss = time.time()
pool = mp.Pool(processes=4)#this computer has 4 cores, I'm only going to use all 4 since I don't need to do anything
results = pool.map(single_run, qlist)
pool.close()
pool.join()
print(results)
ee = time.time()
ee-ss

[232.20171999931335, 235.84447479248047, 233.06574201583862, 223.180645942688, 230.8668520450592, 229.57680892944336, 235.47874188423157, 233.97993993759155, 209.92388916015625, 206.61473202705383]
Querying Data
Querying Data
Querying Data
Querying Data
Finished querying Data in 0.63sFinished querying Data in 0.61sFinished querying Data in 0.61sFinished querying Data in 0.62s



Querying Likelihood
Querying Likelihood
Querying Likelihood
Querying Likelihood
tau = 0.947937011719tau = 0.947937011719tau = 0.942840576172tau = 0.947937011719



158 181
173 184
148 181
153 181
142 158
168 173
138 148
139 153
141 142
174 168
132 138
131 139
149 141
174 174
134 132
135 131
150 149
tau = 0.947937011719
134 134
140 135
150 150
tau = 1.62066650391tau = 0.942840576172
140 140
tau = 0.927551269531

tau = 1.04476928711
tau = 0.947937011719
tau = 0.698211669922
297 311
257 313
tau = 0.703308105469
264 304
288 297
238 257
263 304
241 264
287 288
238 238
248 263
244 241
290 287
tau = 1.04986572266
230 

668.4675850868225

123 123
tau = 0.983612060547
152 149
166 167
tau = 0.917358398438
Finished querying Likelihood in 0.94s152 152
165 166
tau = 1.62066650391

tau = 0.947937011719
165 165
246 314
Querying OneDSearch
tau = 1.62066650391
tau = 0.942840576172
237 246
Finished querying OneDSearch in 49.74s270 311
tau = 1.19256591797
260 237

254 270
291 304
238 260
Querying TwoDSearch
250 254
285 291
232 238
Finished querying TwoDSearch in 182.15s256 250
278 285
224 232

265 256
273 278
224 224
Querying PeakDetect
274 265
269 273
tau = 1.62066650391
Finished querying PeakDetect in 0.51s272 274
267 269
tau = 0.983612060547
272 272
266 267

tau = 1.62066650391
266 266
1622tau = 1.21295166016
tau = 0.881683349609 1736
tau = 1.40661621094
1516
1515 1622
1612 1725
1309 1661
1413 1516
1564 1515
1235 1612
1369 1309
1535 1413
1238 1564
1366 1235
1519 1369
1173 1535
1380 1238
1513 1366
1207 1519
1391 1173
1510 1380
1210 1513
1400 1207
1510 1391
1208 1510
1400 1210
tau = 1.40151977539
 1400
1208Finished querying Likel

In [10]:
ee-ss

668.4675850868225

In [11]:
results

[232.20171999931335,
 235.84447479248047,
 233.06574201583862,
 223.180645942688,
 230.8668520450592,
 229.57680892944336,
 235.47874188423157,
 233.97993993759155,
 209.92388916015625,
 206.61473202705383]

##LINUX 

Since that is so time intensive and we wish to run quite a few of these, we implemented multiprocessing. 

In [4]:
import multiprocessing as mp

In [4]:
qlist = os.listdir("/k2_data/query")
qlist_subset = random.sample(qlist,5)#since there are 158 files in that folder I just wanted to run 5

In [92]:
def single_run(qfile):
    start = time.time()
    #python run.py epicID campaign query_file_path
    #os.system("python run.py " + qfile[4:-2] + " 0 /k2_data/query/" + qfile)
    run.main([qfile[7:-6], 0, "/k2_data/query/" + qfile])
    return time.time() - start

In [28]:
ss = time.time()
pool = mp.Pool(processes=6)#this computer has 8 cores, I'm only going to use 6
results = pool.map(single_run, qlist_subset)
pool.close()
pool.join()
print(results)
ee = time.time()
ee-ss

[544.5501210689545, 584.634486913681, 546.0201680660248, 711.1573400497437, 644.8120350837708]
Querying Data
Querying Data
Querying Data
Querying Data
Querying Data
Finished querying Data in 2.31sFinished querying Data in 2.40sFinished querying Data in 2.43sFinished querying Data in 2.29sFinished querying Data in 2.27s




Querying Likelihood
Querying Likelihood
Querying Likelihood
Querying Likelihood
Querying Likelihood
tau = 0.947937011719tau = 0.947937011719tau = 0.947937011719tau = 0.902069091797tau = 0.942840576172




155 185
173 184
175 182
179 179
140 181
146 155
168 173
169 175
tau = 0.902069091797
127 140
142 146
174 168
171 169
tau = 1.15689086914144 127
137 142
174 174
171 171

142 144
140 137
tau = 0.947937011719
tau = 0.947937011719
306 306
142 142
142 140
tau = 1.62066650391tau = 0.795043945312tau = 1.15689086914
tau = 0.942840576172
142 142


tau = 5.69781494141tau = 0.973419189453tau = 0.947937011719
297 311
300 312


tau = 0.703308105469288 297
296 300
1658249 313

28

-711.1813549995422

The total runtime was 711 seconds or about 11 minutes.

In [31]:
ee-ss

711.1813549995422

Interestingly, each individual run varied more using multiprocessing.

In [30]:
results

[544.5501210689545,
 584.634486913681,
 546.0201680660248,
 711.1573400497437,
 644.8120350837708]

On average, they took 10 minutes each with a standard deviation of about a minute.

In [32]:
np.mean(results)/60

10.103913837273916

In [72]:
np.std(results)

63.865980797904115

This could be because the test was run on a different sample or because of some variable overhead time with multiprocessing.

In [71]:
qq

['EPIC202060911.q',
 'EPIC202073144.q',
 'EPIC202073440.q',
 'EPIC202073353.q',
 'EPIC202093968.q']

#Cacheing

This test the speed of cacheing versus not cacheing for a multiprocessing run. I anticipate the speed will be about the same, maybe slightly slower since data must be written. 

In [130]:
qq

['EPIC202060911.q',
 'EPIC202073144.q',
 'EPIC202073440.q',
 'EPIC202073353.q',
 'EPIC202093968.q']

In [96]:
qlist_subset = qq

In [123]:
reload(sw)
reload(run)

<module 'run' from 'run.py'>

In [124]:
def single_run(qfile):
    start = time.time()
    #python run.py epicID campaign query_file_path
    #os.system("python run.py " + qfile[4:-2] + " 0 /k2_data/query/" + qfile)
    run.main([qfile[7:-6], 0, "/k2_data/query_cache/" + qfile])
    return time.time() - start

In [125]:
ss = time.time()
pool = mp.Pool(processes=6)#this computer has 8 cores, I'm only going to use 6
results = pool.map(single_run, qq)
pool.close()
pool.join()
print(results)
ee = time.time()
ee-ss

[649.0818219184875, 614.6605250835419, 667.3451302051544, 609.237820148468, 640.0954308509827]
Querying Data
Querying Data
Querying Data
Querying Data
Querying Data
Finished querying Data in 1.17sFinished querying Data in 1.16sFinished querying Data in 1.33sFinished querying Data in 1.29sFinished querying Data in 1.16s




Querying LikelihoodQuerying LikelihoodQuerying LikelihoodQuerying LikelihoodQuerying Likelihood




tau = 0.902069091797tau = 0.947937011719tau = 0.947937011719tau = 0.942840576172tau = 0.947937011719




179 179
155 185
173 184
140 181
175 182
tau = 0.902069091797
146 155
168 173
127 140
169 175
tau = 1.15689086914142 146
174 168
144 127
171 169

137 142
174 174
142 144
171 171
306 306
140 137
tau = 0.947937011719
142 142
tau = 0.947937011719
tau = 1.15689086914
142 140
tau = 1.62066650391tau = 0.942840576172
tau = 0.795043945312tau = 5.69781494141142 142

tau = 0.973419189453

tau = 0.947937011719
297 311

300 312
1658tau = 0.703308105469288 297
249 313
296 300
 16

667.5064611434937

The time was slighlty longer with cacheing than without

In [126]:
ee-ss

667.5064611434937

In [127]:
results

[649.0818219184875,
 614.6605250835419,
 667.3451302051544,
 609.237820148468,
 640.0954308509827]

In [128]:
np.mean(results)/60

10.601402427355449